File: bbb/bbb/report/service_record_report/service_record_report.js
// Copyright (c) 2023, invento software limited and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Service Record Report"] = {
  "filters": [
    {
      "fieldname": "from_date",
      "label": __("From Date"),
      "fieldtype": "Date",
      "default": frappe.datetime.get_today(),
      "reqd": 1,
      "width": "60px"
    },
    {
      "fieldname": "to_date",
      "label": __("To Date"),
      "fieldtype": "Date",
      "default": frappe.datetime.get_today(),
      "reqd": 1,
      "width": "60px"
    }, {
      "fieldname": "status",
      "label": __("Status"),
      "fieldtype": "Select",
      "options": ["", "Pending For Service", "In Progress", "Paused", "Submitted", "Cancelled"],
    },
  ],

  "formatter": function (value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);
    if (column.fieldname == "status" && data && data.status === "Pending For Service") {
      value = "<span style='color:orange'>" + value + "</span>";
    } else if (column.fieldname == "status" && data && data.status === "In Progress") {
      value = "<span style='color:green'>" + value + "</span>";
    } else if (column.fieldname == "status" && data && data.status === "Paused") {
      value = "<span style='color:darkgrey'>" + value + "</span>";
    } else if (column.fieldname == "status" && data && data.status === "Submitted") {
      value = "<span style='color:blue'>" + value + "</span>";
    } else if (column.fieldname == "status" && data && data.status === "Cancelled") {
      value = "<span style='color:red'>" + value + "</span>";
    }


    if (column.fieldname == 'view') {
      value = `<div style="background-color: #f5f7fa; cursor: pointer;" class="disabled" onclick="(()=> (frappe.query_reports['Service Record Report']).set_route('` + data.name + `'))()">View</div>`;
    }
    return value;
  },
  "set_route":(value)=>{
    window.open("/app/service-record/"+ value, "_blank")
  }
};

File: bbb/bbb/report/service_record_report/service_record_report.py
# Copyright (c) 2022, invento software limited and contributors
# For license information, please see license.txt

import frappe
from frappe import _


def execute(filters=None):
    columns, data = get_columns(), get_service_data(filters)
    return columns, data


def get_columns():
    """return columns"""
    columns = [
        {"label": _("View"), "fieldname": "view", "fieldtype": "Text", "width": 80},
        {"label": _("Service Name"), "fieldname": "service_name", "fieldtype": "Text", "width": 600},
        {"label": _("Service Person 1"), "fieldname": "service_person_1", "fieldtype": "Text", "width": 150},
        {"label": _("Service Person 2"), "fieldname": "service_person_2", "fieldtype": "Text", "width": 150},
        {"label": _("Service Person 3"), "fieldname": "service_person_3", "fieldtype": "Text", "width": 150},
        {"label": _("Service Person 3"), "fieldname": "service_person_4", "fieldtype": "Text", "width": 150},
        {"label": _("Status"), "fieldname": "status", "fieldtype": "Text", "width": 150},
    ]
    return columns


def get_conditions(filters):

    conditions = []
    if filters.get("from_date"):
        conditions.append("service_record.service_date >= '%s'" % filters.get("from_date"))

    if filters.get("to_date"):
        conditions.append("service_record.service_date <= '%s'" % filters.get("to_date"))

    if filters.get("status"):
        conditions.append("service_record.status = '%s'" % filters.get("status"))

    if conditions:
        conditions = " and ".join(conditions)

    return conditions


def get_service_data(filters):
    conditions = get_conditions(filters)
    query_result = frappe.db.sql("""
    SELECT name,
    service_name,
    service_person_1,
    service_person_2,
    service_person_3,
    service_person_4,
    status
    FROM `tabService Record` service_record
    WHERE %s""" % (conditions), as_dict=1)

    return query_result
